### ISSUE TITLE
IDE threading change: Don't block during initialization

### ISSUE THREAD

**Issue by jvilk-stripe** [sorbet/sorbet#5010]
### Problem

If we want to support a package-local Sorbet IDE experience,  then the language server needs to support interruptions during initialization.

Sorbet’s language server assumes that there is a single GlobalState, and the responsibility for maintaining this GlobalState is spread across two threads:
* An _indexer + coordinator_ thread that decides if a file update can be typechecked without retypechecking the whole project. It then schedules typechecking work (and preempts slow operations to process fast operations to improve latency).
* A _typechecking_ thread that owns a WorkerPool and actually runs typechecking and other IDE operations.

There is one point of awkwardness in this design: During initialization, the indexer has to borrow the typechecking thread’s WorkerPool to efficiently parse and hash every file in pay-server so that it can make efficient typechecking decisions. During this process, Sorbet becomes unresponsive and does not respond to IDE requests, which is OK right now because initialization only happens once -- when Sorbet starts up -- and there are no meaningful requests that could possibly be processed until that activity completes.

In a packager world, initialization would happen whenever the user opens a file in a package that isn’t already initialized. It will be a common activity. While this is happening, Sorbet should be able to interrupt that process and answer queries about other packages.

### Proposed solution

We should change the threading model so that the indexer can request that the typechecker complete work (in this case, parsing every file in pay-server) and the typechecker can pass the result back to the indexer. I envision this working as follows:
* The coordinator passes the task queue to the indexer, and asks for the next task to process.
* The indexer sees that it needs to initialize, and generate an initialization task and returns it.
* The coordinator schedules that work on the typechecker thread.
* The typechecker completes the work and enqueues a new task to the task queue containing the data the indexer needs.
* The coordinator wakes up, passes the task queue to the indexer, and the indexer pulls off the data it requires.

While initialization is happening, the indexer can select which tasks in the queue can be processed. If all of them are blocked on initialization, it can tell the coordinator that fact and the coordinator will sleep until the next time the task queue is mutated.

Note that these changes need to work in the single-threaded emscripten loop. In the above design, I expect that the task queue will be passed through to typechecking, and that, if there is anything in it, the emscripten loop will pass it through to indexer itself.
